bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java

```diff
@@ -1386,6 +1391,31 @@ private static boolean checkIsRecordType(MapValue sourceRecordValue, BRecordType
         return true;
     }
 
+    private static boolean containsInvalidNeverField(Field sourceField, BRecordType sourceRecordType) {
+        if (sourceField != null) {
+            return !containsNeverType(sourceField.getFieldType());
+        }
+        if (sourceRecordType.isSealed()) {
+            return true;
+        }
+        return !containsNeverType(sourceRecordType.getRestFieldType());
+    }
+
+    private static boolean containsNeverType(Type fieldType) {
+        int fieldTag = fieldType.getTag();
+        if (fieldTag == TypeTags.NEVER_TAG) {
+            return true;
+        }
+        if (fieldTag == TypeTags.UNION_TAG) {
+            for (Type member : ((BUnionType) fieldType).getOriginalMemberTypes()) {
```

Comment:
Shall we move this to seperate variable ((BUnionType) fieldType).getOriginalMemberTypes()